lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+  matchAndRewrite(ArrayGetOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto inputType = type_cast<ArrayType>(op.input().getType());
+    Type targetIndexType = IntegerType::get(
+        getContext(),
+        inputType.getSize() == 1 ? 1 : llvm::Log2_64_Ceil(inputType.getSize()));
+
+    if (op.index().getType().getIntOrFloatBitWidth() ==
+        targetIndexType.getIntOrFloatBitWidth())
+      return failure(); // nothing to do
+
+    // Narrow the index value.
+    auto narrowedIndex =
+        narrowValueToArrayWidth(rewriter, op.input(), op.index());
+    rewriter.replaceOpWithNewOp<ArrayGetOp>(op, op.input(), narrowedIndex);
+    return success();
+  }
+};
+
+// Generic pattern to convert parametric result types.
+struct ParametricTypeConversionPattern : public ConversionPattern {
+  ParametricTypeConversionPattern(MLIRContext *ctx,
+                                  TypeConverter &typeConverter,
+                                  ArrayAttr parameters)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), /*benefit=*/1,
+                          ctx),
+        parameters(parameters) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<Value, 4> convertedOperands;
+    // Update the result types of the operation
+    bool ok = true;
+    rewriter.updateRootInPlace(op, [&]() {
+      // Mutate result types
+      for (auto &it : llvm::enumerate(op->getResultTypes())) {
+        auto res = evaluateParametricType(op->getLoc(), parameters, it.value());
+        ok &= succeeded(res);
+        if (!ok)
+          return;
+        op->getResult(it.index()).setType(res.getValue());
+      }
+    });
+
+    // Note: 'operands' have already been converted with the supplied type
+    // converter to this pattern. Make sure that we materialize this conversion
+    // by updating the operands to op.
+    op->setOperands(operands);
+
+    return success(ok);
+  };
+  ArrayAttr parameters;
+};
+
+struct HWSpecializePass : public sv::HWSpecializeBase<HWSpecializePass> {
+  void runOnOperation() override;
+};
+
+static void populateTypeConversion(Location loc, TypeConverter &typeConverter,
+                                   ArrayAttr parameters) {
+  // Possibly parametric types
+  typeConverter.addConversion([=](hw::IntType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+  typeConverter.addConversion([=](hw::ArrayType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+
+  // Valid target types.
+  typeConverter.addConversion([](mlir::IntegerType type) { return type; });
+}
+
+// Specializes the provided 'base' module into the 'target' module. By doing
+// so, we create a new module which
+// 1. has no parameters
+// 2. has a name composing the name of 'base' as well as the 'parameters'
```

Comment:
can you be more precise about "composing" the name? As far as I can tell the current algorithm is something like:

`originalModuleName_fooIntParameter_5_barStringParameter_Hello`

how are parameter names escaped if they are not immediately verilog legal?

Reply:
> how are parameter names escaped if they are not immediately verilog legal?

CIRCT handles that just before ExportVerilog... not to my liking for certain characters, but it'll be valid.